### PR TITLE
RepositoryServer CR update validation

### PR DESCRIPTION
**Pull request by r4rajat**
## Change Overview

Following fields in RepositoryServer CRD need to be made immutable to control the CR update event.

- .spec.storage.secretRef
- .spec.repository.rootPath
- .spec.server.adminSecretRef
- .spec.server.tlsSecretRef

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [x] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [x] :robot: Test

## Issues 

- fixes #issue-number

## Test Plan

- [x] :muscle: Manual
- [x] :zap: Unit test
- [ ] :green_heart: E2E

#### 1) Install Kanister using Helm

```bash
helm install kanister ./helm/kanister-operator \
--namespace kanister \
--set image.repository=r4rajat/controller \
--set image.tag=v5 \
--set repositoryServerControllerImage.registry=r4rajat \
--set repositoryServerControllerImage.name=repo-server-controller \
--set repositoryServerControllerImage.tag=v5 \
--set controller.parallelism=10 \
--set repositoryServerController.enabled=true \
--create-namespace
```

#### 2) Creating RepositoryServer CR

```bash
kubectl apply -f - <<EOF
---
apiVersion: cr.kanister.io/v1alpha1
kind: RepositoryServer
metadata:
  labels:
    app.kubernetes.io/name: repositoryserver
    app.kubernetes.io/instance: repositoryserver-sample
    app.kubernetes.io/part-of: kanister
    app.kuberentes.io/managed-by: kustomize
    app.kubernetes.io/created-by: kanister
  name: kopia-repo-server-1
  # namespace: kanister
spec:
  storage:
    secretRef:
      name: s3-location
      namespace: kanister
    credentialSecretRef:
      name: s3-creds
      namespace: kanister
  repository:
    rootPath: /repo-controller/
    passwordSecretRef:
      name: repo-pass 
      namespace: kanister
    username: kanisteradmin
    hostname: time-log.app
  server:
    adminSecretRef:
      name: repository-admin-user
      namespace: kanister
    tlsSecretRef:
      name: repository-server-tls-cert
      namespace: kanister
    userAccess:
      userAccessSecretRef:
        name: repository-server-user-access
        namespace: kanister
      username: kanisteruser
EOF

repositoryserver.cr.kanister.io/kopia-repo-server-1 created
```

#### 3) Try to update rootPath

```bash
kubectl apply -f - <<EOF
---
apiVersion: cr.kanister.io/v1alpha1
kind: RepositoryServer
metadata:
  labels:
    app.kubernetes.io/name: repositoryserver
    app.kubernetes.io/instance: repositoryserver-sample
    app.kubernetes.io/part-of: kanister
    app.kuberentes.io/managed-by: kustomize
    app.kubernetes.io/created-by: kanister
  name: kopia-repo-server-1
  # namespace: kanister
spec:
  storage:
    secretRef:
      name: s3-location
      namespace: kanister
    credentialSecretRef:
      name: s3-creds
      namespace: kanister
  repository:
    rootPath: /repo-controller/update-test/
    passwordSecretRef:
      name: repo-pass 
      namespace: kanister
    username: kanisteradmin
    hostname: time-log.app
  server:
    adminSecretRef:
      name: repository-admin-user
      namespace: kanister
    tlsSecretRef:
      name: repository-server-tls-cert
      namespace: kanister
    userAccess:
      userAccessSecretRef:
        name: repository-server-user-access
        namespace: kanister
      username: kanisteruser
EOF

The RepositoryServer "kopia-repo-server-1" is invalid: spec.repository.rootPath: Invalid value: "string": Value is immutable
```

### CODE REVIEW

**Comment by kale-amruta** [comment]
can we add a unit test for this? @r4rajat

File: pkg/customresource/repositoryserver.yaml [hunk 1]
```diff
@@ -11,289 +11,319 @@ spec:
...
                       https://kopia.io/docs/reference/command-line/common/#commands-to-manipulate-repository'
-                    type: string
-                  username:
-                    description: If specified, these values will be used by the controller
-                      to override default username when connecting to the repository
-                      from the server.
-                    type: string
-                required:
-                - passwordSecretRef
-                - rootPath
```
Comment (kale-amruta) [inline pkg/customresource/repositoryserver.yaml:82]: why are required properties being removed?

**Review by PrasadG193 — COMMENTED** [review]
LGTM

File: pkg/customresource/repositoryserver.yaml [hunk 2]
```diff
@@ -11,289 +11,319 @@ spec:
...
+                            maxLength: 316
+                            pattern: ^([a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*/)?(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])$
+                            type: string
+                        required:
+                          - lastTransitionTime
+                          - message
+                          - reason
+                          - status
+                          - type
+                        type: object
```
Comment (kale-amruta) [inline pkg/customresource/repositoryserver.yaml:294]: Why are we making this required?

**Comment by r4rajat** [comment]
> can we add a unit test for this? @r4rajat

> Done @kale-amruta

File: pkg/apis/cr/v1alpha1/repositoryserver_types_test.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	c.Assert(rs.Spec.Repository.PasswordSecretRef.Namespace, Equals, "kanister")
+	c.Assert(rs.Spec.Repository.Username, Equals, "test-repository-user")
+	c.Assert(rs.Spec.Repository.Hostname, Equals, "localhost")
+	c.Assert(rs.Spec.Server.AdminSecretRef.Name, Equals, "test-repository-admin-user")
+	c.Assert(rs.Spec.Server.AdminSecretRef.Namespace, Equals, "kanister")
+	c.Assert(rs.Spec.Server.TLSSecretRef.Name, Equals, "test-repository-server-tls-cert")
+	c.Assert(rs.Spec.Server.TLSSecretRef.Namespace, Equals, "kanister")
+	c.Assert(rs.Spec.Server.UserAccess.UserAccessSecretRef.Name, Equals, "test-repository-server-user-access")
+	c.Assert(rs.Spec.Server.UserAccess.UserAccessSecretRef.Namespace, Equals, "kanister")
+	c.Assert(rs.Spec.Server.UserAccess.Username, Equals, "test-kanister-user")
```
Comment (kale-amruta) [inline pkg/apis/cr/v1alpha1/repositoryserver_types_test.go:82]: I thought we are testing the validation here, but taking a closer look, I think we arent. Can we add create and apply update to the CR and see if we get an error?

File: pkg/apis/cr/v1alpha1/repositoryserver_types.go [hunk 4]
```diff
@@ -51,20 +51,26 @@ type RepositoryServerSpec struct {
...
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"
 	SecretRef corev1.SecretReference `json:"secretRef"`
 	// CredentialSecretRef stores the credentials required
 	// to connect to the object storage specified in `SecretRef` field
 	CredentialSecretRef corev1.SecretReference `json:"credentialSecretRef"`
 }
 
 // Repository has the details required by the repository server to connect to kopia repository
+// +kubebuilder:validation:XValidation:rule="!has(oldSelf.rootPath) || has(self.rootPath)",message="rootPath is required once set"
```
[the remark this answers is not in the document]
Comment (r4rajat) [reply pkg/apis/cr/v1alpha1/repositoryserver_types.go:67]: it means `Once set, field must not be allowed to be removed`

[the remark this answers is not in the document]
Comment (r4rajat) [reply pkg/apis/cr/v1alpha1/repositoryserver_types.go:67]: I'll update the Message

File: pkg/apis/cr/v1alpha1/repositoryserver_types_test.go (hunk 3 above)
Comment (r4rajat) [reply pkg/apis/cr/v1alpha1/repositoryserver_types_test.go:82]: I have added the tests in https://github.com/kanisterio/kanister/pull/1997/files#diff-2ea977fa45058d073275fccb79a53b9f41907d3b9e8d1f9488617c3794544b67

File: pkg/apis/cr/v1alpha1/repositoryserver_types_test.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      userAccessSecretRef:
+        name: test-repository-server-user-access
+        namespace: kanister
+      username: test-kanister-user
+`
+
+func TestRepositoryServer(t *testing.T) { TestingT(t) }
+
+func (s *TypesSuite) TestRepositoryServerDecode(c *C) {
+	rs, err := getRepositoryServerFromSpec([]byte(repoServerSpec))
```
[the remark this answers is not in the document]
Comment (r4rajat) [reply pkg/apis/cr/v1alpha1/repositoryserver_types_test.go:64]: Added in https://github.com/kanisterio/kanister/pull/1997/files#diff-2ea977fa45058d073275fccb79a53b9f41907d3b9e8d1f9488617c3794544b67

**Review by kale-amruta — APPROVED** [review]
lgtm

**Pull request merged**